python-checks/src/main/java/org/sonar/python/checks/SklearnPipelineParameterAreCorrectCheck.java

```diff
@@ -0,0 +1,182 @@
+            .ifPresent(location -> subscriptionContext.addIssue(location.tree, MESSAGE));
+        }
+      });
+    }
+  }
+
+  private static Optional<FunctionSymbol> getInitFunctionSymbol(ClassSymbol classSymbol) {
+    return classSymbol.declaredMembers().stream().filter(
+      memberSymbol -> "__init__".equals(memberSymbol.name())).findFirst().map(FunctionSymbol.class::cast);
+  }
+
+  private static Stream<Expression> getExpressionsFromArgument(@Nullable RegularArgument argument) {
+    return Optional.ofNullable(argument)
+      .map(RegularArgument::expression)
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(ListLiteralImpl.class))
+      .map(ListLiteralImpl::elements)
+      .map(ExpressionList::expressions)
+      .stream()
+      .flatMap(Collection::stream);
+  }
+
+  private static Map<String, ClassSymbol> parsePipeline(CallExpression callExpression) {
+    var stepsArgument = TreeUtils.nthArgumentOrKeyword(0, "steps", callExpression.arguments());
+    var out = new HashMap<String, ClassSymbol>();
+
+    getExpressionsFromArgument(stepsArgument)
+      .map(
+        TreeUtils.toInstanceOfMapper(TupleImpl.class))
+      .filter(Objects::nonNull)
+      .map(TupleImpl::elements)
+      .filter(elements -> elements.size() == 2)
+      .forEach(
+        tuple -> getResult(tuple).ifPresent(result1 -> out.put(result1.stepName(), result1.classifierName())));
+    return out;
+  }
+
+  private static Optional<Result> getResult(List<Expression> tuple) {
+    var step = tuple.get(0);
+    var classifier = tuple.get(1);
+    var stepName = Optional.ofNullable(step)
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(StringLiteral.class))
+      .map(StringLiteral::trimmedQuotesValue);
+    var classifierName = Optional.ofNullable(classifier)
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(CallExpression.class))
+      .map(CallExpression::calleeSymbol)
+      .map(ClassSymbol.class::cast);
+
+    return stepName.flatMap(stepName1 -> classifierName.map(classifierName1 -> new Result(stepName1, classifierName1)));
+  }
+
+  private record Result(String stepName, ClassSymbol classifierName) {
+  }
+
+  private static Map<String, Set<StringAndTree>> getStepAndParametersFromArguments(CallExpression callExpression) {
+    return callExpression.arguments().stream().filter(argument -> argument.is(Tree.Kind.REGULAR_ARGUMENT))
+      .map(RegularArgument.class::cast)
+      .map(RegularArgument::keywordArgument)
+      .filter(Objects::nonNull)
+      .map(SklearnPipelineParameterAreCorrectCheck::getStepAndParameterFromName)
+      .filter(Optional::isPresent)
+      .map(Optional::get)
+      .collect(Collectors.toMap(StepAndParameter::step, stepAndParameter -> Set.of(new StringAndTree(stepAndParameter.parameter(), stepAndParameter.location())),
+        (set1, set2) -> {
+          var set = new HashSet<>(set1);
+          set.addAll(set2);
+          return set;
+        }));
```

Comment:
Maybe here it would be a good idea to move the collector to another function. it would help a bit with the comprehension of the code.